app/lib/analyzer/backend.dart

```diff
@@ -110,33 +111,47 @@ class AnalysisBackend {
           .append(Package, id: packageName)
           .append(PackageVersion, id: packageVersion)
     ]);
-    // package or version does not exists
-    if (versions.first == null) return false;
+    // Does package and version exist?
+    final PackageVersion pv = versions.single;
+    if (pv == null) return false;
 
+    // Does package have any analysis?
     final List<PackageAnalysis> list =
         await db.lookup([db.emptyKey.append(PackageAnalysis, id: packageName)]);
-    // no analysis for the package
-    if (list.first == null) return true;
-
     final PackageAnalysis packageAnalysis = list.first;
+    if (packageAnalysis == null) return true;
+
+    // Does package have newer version than latest analyzed version?
+    if (isNewer(packageAnalysis.latestSemanticVersion, pv.semanticVersion)) {
+      return true;
+    }
+
+    // Does package have analysis for the current version?
     final PackageVersionAnalysis versionAnalysis =
         (await db.lookup([packageAnalysis.latestVersionKey])).first;
+    if (versionAnalysis == null) return true;
 
-    if (versionAnalysis.analysisVersion == analysisVersion) {
-      final DateTime now = new DateTime.now().toUtc();
-      if (now.difference(versionAnalysis.analysisTimestamp).inHours < 4) {
-        // latest analysis has the same version and happened in the past 4 hours
-        return false;
-      }
-    } else {
-      final Version analysisSemantic = new Version.parse(analysisVersion);
-      if (isNewer(analysisSemantic, versionAnalysis.semanticAnalysisVersion)) {
-        // existing analysis version is newer, current analyzer is obsolete?
-        // TODO: turn off task polling on this instance
-        _logger.warning(
-            'Newer analysis version detected, current instance is obsolete.');
-        return false;
-      }
+    // Is current analysis version newer?
+    final Version analysisSemantic = new Version.parse(analysisVersion);
+    if (isNewer(versionAnalysis.semanticAnalysisVersion, analysisSemantic)) {
+      return true;
+    }
+
+    // Is current analysis version obsolete?
+    if (isNewer(analysisSemantic, versionAnalysis.semanticAnalysisVersion)) {
+      // existing analysis version is newer, current analyzer is obsolete?
+      // TODO: turn off task polling on this instance
+      _logger.warning(
+          'Newer analysis version detected, current instance is obsolete.');
+      return false;
+    }
+
+    assert(versionAnalysis.analysisVersion == analysisVersion);
+
+    // Is latest analysis older than 4 hours?
+    final DateTime now = new DateTime.now().toUtc();
+    if (now.difference(versionAnalysis.analysisTimestamp).inHours < 4) {
```

Comment:
So I've looked up the constant we should share, namely the one defined [here](https://github.com/dart-lang/pub-dartlang-dart/blob/fb5f9479db5233fdea5ae4fa6da48308e5609b20/app/lib/analyzer/task_sources.dart#L111):

```dart
  ....
  await new Future.delayed(const Duration(days: 1));
```

This duration has precisely the meaning to not re-analyze historical data more often than once a day. The duration in this CL right here should have the same meaning, so we should put it into a shared place and use the constant in both places.